src/GafferScene/CameraQuery.cpp

```diff
@@ -0,0 +1,551 @@
+}
+
+const Gaffer::StringPlug *CameraQuery::locationPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 2 );
+}
+
+Gaffer::ArrayPlug *CameraQuery::queriesPlug()
+{
+	return getChild<ArrayPlug>( g_firstPlugIndex + 3 );
+}
+
+const Gaffer::ArrayPlug *CameraQuery::queriesPlug() const
+{
+	return getChild<ArrayPlug>( g_firstPlugIndex + 3 );
+}
+
+Gaffer::ArrayPlug *CameraQuery::outPlug()
+{
+	return getChild<ArrayPlug>( g_firstPlugIndex + 4 );
+}
+
+const Gaffer::ArrayPlug *CameraQuery::outPlug() const
+{
+	return getChild<ArrayPlug>( g_firstPlugIndex + 4 );
+}
+
+Gaffer::AtomicCompoundDataPlug *CameraQuery::internalParametersPlug()
+{
+	return getChild<AtomicCompoundDataPlug>( g_firstPlugIndex + 5 );
+}
+
+const Gaffer::AtomicCompoundDataPlug *CameraQuery::internalParametersPlug() const
+{
+	return getChild<AtomicCompoundDataPlug>( g_firstPlugIndex + 5 );
+}
+
+Gaffer::StringPlug *CameraQuery::addQuery( const Gaffer::ValuePlug *plug, const std::string &parameter )
+{
+	StringPlugPtr childQueryPlug = new StringPlug( "query0" );
+	childQueryPlug->setValue( parameter );
+
+	ValuePlugPtr newOutPlug = new ValuePlug( "out0", Plug::Out );
+	newOutPlug->addChild( new StringPlug( "source", Plug::Out ) );
+	newOutPlug->addChild( plug->createCounterpart( "value", Plug::Out ) );
+
+	outPlug()->addChild( newOutPlug );
+	queriesPlug()->addChild( childQueryPlug );
+
+	return childQueryPlug.get();
+}
+
+void CameraQuery::removeQuery( Gaffer::StringPlug *plug )
+{
+	const ValuePlug *oPlug = outPlugFromQuery( plug );
+
+	queriesPlug()->removeChild( plug );
+	outPlug()->removeChild( const_cast<ValuePlug *>( oPlug ) );
+}
+
+void CameraQuery::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	ComputeNode::affects( input, outputs );
+
+	if( input == cameraModePlug() ||
+		( input == locationPlug() && cameraModePlug()->isSetToDefault() ) ||
+		input == scenePlug()->existsPlug() ||
+		input == scenePlug()->objectPlug() ||
+		input == scenePlug()->globalsPlug()
+	)
+	{
+		outputs.push_back( internalParametersPlug() );
+	}
+	else if( input == internalParametersPlug() )
+	{
+		addChildPlugsToAffectedOutputs( outPlug(), outputs );
+	}
+	else if( queriesPlug()->isAncestorOf( input ) )
+	{
+		const StringPlug *queryPlug = runTimeCast<const StringPlug>( input );
+		if( queryPlug == nullptr )
+		{
+			throw IECore::Exception( "CameraQuery::affects : Query plugs must be \"StringPlug\"" );
+		}
+
+		const ValuePlug *valuePlug = valuePlugFromQuery( queryPlug );
+
+		addChildPlugsToAffectedOutputs( valuePlug, outputs );
+
+		outputs.push_back( sourcePlugFromQuery( queryPlug ) );
+	}
+}
+
+void CameraQuery::hash( const Gaffer::ValuePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	if( output == internalParametersPlug() )
+	{
+		std::string location = "";
+		// The globals affect the camera parameters via SceneAlgo::applyCameraGlobals()
+		h.append( scenePlug()->globalsHash() );
```

Comment:
This might be more self explanatory if it was moved into the `scenePlug()-existsPlug()` block below, so it's in the same spot we actually call `applyCameraGlobals()`.